### PR TITLE
DMI standardproblem

### PR DESCRIPTION
**Pull request by DiegoDeGusem**
I have created the 1D, 2D and 3D DMI standarproblems from [here](https://iopscience.iop.org/article/10.1088/1367-2630/aaea1c). For each problem I initialize the system and let it minimize. Then I compare the magnetization obtained in mumax⁺ by the one obtained in mumax³.

### CODE REVIEW

File: test/mumax3/test_mumax3_DMI_standardproblem3D.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    magnet = Ferromagnet(world, Grid(gridsize), geometry=geo)
+
+    magnet.enable_demag = False
+    magnet.msat = Ms
+    magnet.aex = A
+    magnet.dmi_tensor.set_bulk_dmi(D)
+
+    magnet.bias_magnetic_field = (0,0,Bz)
+
+    magnet.magnetization = blochskyrmion(magnet.center, 91.5e-9, charge, pol)
```
Comment (ilateur) [inline test/mumax3/test_mumax3_DMI_standardproblem3D.py:48]: changing the radius from 91.5e-9 to about 18e-9 makes it so the test "only" takes 14 seconds for me

File: test/mumax3/test_mumax3_DMI_standardproblem3D.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+            dx := lx / nx
+            dy := ly / ny
+            dz := lz / nz
+
+            SetGridSize(nx, ny, nz)
+            SetCellSize(dx, dy, dz)
+
+            // Define the cylinder
+            SetGeom(Circle(lx))
```
Comment (ilateur) [inline test/mumax3/test_mumax3_DMI_standardproblem3D.py:72]:
```python
f"""
    SetGridSize{gridsize}
    SetCellSize{cellsize}
    SetGeom(Circle({diam}))
"""
```

File: test/mumax3/test_mumax3_DMI_standardproblem3D.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+            // External field in T
+            B_ext = vector(0, 0, {Bz})
+
+            // No Demag
+            NoDemagSpins = 1
+
+            m = BlochSkyrmion(1, -1)
+
+            OutputFormat = OVF2_TEXT
```
Comment (ilateur) [inline test/mumax3/test_mumax3_DMI_standardproblem3D.py:86]: not needed?

File: test/mumax3/test_mumax3_DMI_standardproblem2D.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            Ku1         = {Ku}
+            anisU       = vector{tuple(anisU)}
+            Dind        = {D}
+            // No Demag:
+            NoDemagSpins = 1
+
+            // Initial state
+            m = Neelskyrmion({charge}, {pol})
+
+            OutputFormat = OVF2_TEXT
```
Comment (ilateur) [inline test/mumax3/test_mumax3_DMI_standardproblem2D.py:86]: not needed?

File: test/mumax3/test_mumax3_DMI_standardproblem3D.py
Comment (ilateur) [inline test/mumax3/test_mumax3_DMI_standardproblem3D.py:1]: We can take the opportunity again to test both open and closed BC.

**Review by ilateur — CHANGES_REQUESTED** [review]
nice that these exist :)

File: test/mumax3/test_mumax3_DMI_standardproblem1D.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        f"""
+            setcellsize{cellsize}
+            setgridsize{gridsize}
+            msat = {Ms}
+            aex = {A}
+            Ku1 = {Ku}
+            anisU = vector{anisU}
+            {DMI} = {D}
+            enabledemag = False
+            SetPBC(0,2,0)
```
[the remark this answers is not in the document]
Comment (ilateur) [reply test/mumax3/test_mumax3_DMI_standardproblem1D.py:67]: Comparison of 3 and + for Dbulk with closed BC.
image

[the remark this answers is not in the document]
Comment (ilateur) [reply test/mumax3/test_mumax3_DMI_standardproblem1D.py:67]: Ah it looks like mumax+ only has open BC. @lamoreel can you confirm or deny?

File: test/mumax3/test_mumax3_DMI_standardproblem2D.py
[the remark this answers is not in the document]
Comment (ilateur) [reply test/mumax3/test_mumax3_DMI_standardproblem2D.py:1]: or maybe not

File: test/mumax3/test_mumax3_DMI_standardproblem3D.py
Comment (ilateur) [reply test/mumax3/test_mumax3_DMI_standardproblem3D.py:1]: or maybe not

File: test/mumax3/test_mumax3_DMI_standardproblem1D.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lamoreel) [reply test/mumax3/test_mumax3_DMI_standardproblem1D.py:67]: As of now we provide (for magnetodynamic simulations) open, periodic and Neumann boundary conditions. I have no idea what you mean by "closed" :).

[the remark this answers is not in the document]
Comment (lamoreel) [reply test/mumax3/test_mumax3_DMI_standardproblem1D.py:67]: Neumann boundaries are implemented for every type of exchange interaction (including DMI) and these _should_ be correct... (but this is what tests are for :)).

File: test/mumax3/test_mumax3_DMI_standardproblem1D.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    magnet.enable_openbc = True
+
+    magnet.msat = Ms
+    magnet.aex = A
+    magnet.ku1 = Ku
+    magnet.anisU = anisU
+
+    if interfacial:
+        DMI = "Dind"
+        D = -D
```
[the remark this answers is not in the document]
Comment (DiegoDeGusem) [reply test/mumax3/test_mumax3_DMI_standardproblem1D.py:47]: I thought so too. But when I checked the paper and their code they also did this in mumax³ while the other simulators used +D

File: test/mumax3/test_mumax3_DMI_standardproblem2D.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+@pytest.mark.mumax3
+class TestDMI2D:
+    """Compare the results of the simulations by comparing the magnetizations.
+    """
+
+    def test(self):
+        magnet, mumax3sim = simulations()
+        err = max_relative_error(magnet.magnetization.eval(), mumax3sim.get_field("m"))
+        assert err < ATOL
```
[the remark this answers is not in the document]
Comment (DiegoDeGusem) [reply test/mumax3/test_mumax3_DMI_standardproblem2D.py:104]: If we also want to test multiple BC I would keep as a class with then 2 functions. It could also be a `pytest.fixture` here, but then I don't know if that is still possible in the 1D case because the function already has a `fixture` :thinking:. There are also a lot of other tests that use this build

File: test/mumax3/test_mumax3_DMI_standardproblem1D.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (DiegoDeGusem) [reply test/mumax3/test_mumax3_DMI_standardproblem1D.py:67]: It works :smile: . The issue lies with mumax³. If the magnet does not lie along the z-direction they ignore that component in the DMI (which seems wrong). I have "fixed" this by rotating the magnet so it lies along the z-direction.

**Comment by DiegoDeGusem** [comment]
I have added 3 more tests for AFM. They still use the same setup as the 3 standard problems. The AFM exchange parameters are set to 0 and both sublattices are initialized in the same way. The magnetization of the AFM is then compared to the FM result.

**Comment by lamoreel** [comment]
I'm not sure how useful these AFM versions are as tests. They don't simulate physical systems and only test part of the used kernel functions. For us it's good to know that they work, but I think we need more formal/complete tests in this case.

**Comment by ilateur** [comment]
They don't test enough, but if it's good for us to know, then it's good that they exist :)

**Comment by lamoreel** [comment]
Yes. Perhaps these can be removed later on when we develop more general ones.

File: test/mumax3/test_mumax3_DMI_standardproblem1D.py [hunk 8]
```diff
@@ -1,5 +1,8 @@
 """This test is based on the 1D case in
-   https://iopscience.iop.org/article/10.1088/1367-2630/aaea1c"""
+   https://iopscience.iop.org/article/10.1088/1367-2630/aaea1c
+   The magnet lies along the z-direction, because mumax³ is not correct. They ignore
```
Comment (lamoreel) [inline test/mumax3/test_mumax3_DMI_standardproblem1D.py:3]: "mumax3 is not correct" is a very bold statement. I would be more specific and refer to the issue number in its repo instead.

File: test/test_DMI_1D_afm.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
+"""This test is based on the 3D case in
```
Comment (lamoreel) [inline test/test_DMI_1D_afm.py:1]: 1D, not 3D.

File: test/test_DMI_1D_afm.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    magnet_AFM.afmex_nn = 0
+    magnet_AFM.ku1 = Ku
+    magnet_AFM.anisU = anisU
+
+    if interfacial:
+        for sub in magnet_AFM.sublattices:
+            sub.dmi_tensor.set_interfacial_dmi(D)
+    else:
+        for sub in magnet_AFM.sublattices:
+            sub.dmi_tensor.set_bulk_dmi(D)
```
Comment (lamoreel) [inline test/test_DMI_1D_afm.py:68]: Please move the `for` loop one level higher because `for` loops scare me so I only want one :).

File: test/test_DMI_2D_afm.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
+"""This test is based on the 3D case in
```
Comment (lamoreel) [inline test/test_DMI_2D_afm.py:1]: 1D.

File: test/test_DMI_2D_afm.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+
+ATOL = 1e-4
+def max_absolute_error(result, wanted):
+    err = np.linalg.norm(result - wanted, axis=0)
+    return np.max(err)
+
+def simulations(openBC):
+    """This simulates a 2D circle with interfacial DMI and a Neél skyrmion
+       a ferromagnet and an antiferromagnet."""
```
Comment (lamoreel) [inline test/test_DMI_2D_afm.py:20]: This sentence is missing a/some word(s).

File: test/test_DMI_2D_afm.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    magnet_AFM.aex = A
+    magnet_AFM.afmex_cell = 0
+    magnet_AFM.afmex_nn = 0
+    magnet_AFM.ku1 = Ku
+    magnet_AFM.anisU = anisU
+
+    for sub in magnet_AFM.sublattices:
+        sub.dmi_tensor.set_interfacial_dmi(D)
+    magnet_AFM.sub1.magnetization = neelskyrmion(magnet.center, skyrmion_radius, charge, pol)
+    magnet_AFM.sub2.magnetization = neelskyrmion(magnet.center, skyrmion_radius, charge, pol)
```
Comment (lamoreel) [inline test/test_DMI_2D_afm.py:78]: Can't this be set in one line? (`magnet_AFM.magnetization = ...`)

File: test/test_DMI_3D_afm.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+
+ATOL = 1e-4
+def max_absolute_error(result, wanted):
+    err = np.linalg.norm(result - wanted, axis=0)
+    return np.max(err)
+
+def simulations(openBC):
+    """This simulates a 3D cylinder with bulk DMI and a bloch skyrmion
+       a ferromagnet and an antiferromagnet."""
```
Comment (lamoreel) [inline test/test_DMI_3D_afm.py:20]: Same comment.

File: test/test_DMI_3D_afm.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    magnet_AFM.aex = A
+    magnet_AFM.afmex_cell = 0
+    magnet_AFM.afmex_nn = 0
+
+    for sub in magnet_AFM.sublattices:
+        sub.dmi_tensor.set_bulk_dmi(D)
+
+    magnet_AFM.bias_magnetic_field = (0,0,Bz)
+
+    magnet_AFM.sub1.magnetization = blochskyrmion(magnet.center, skyrmion_radius, charge, pol)
```
Comment (lamoreel) [inline test/test_DMI_3D_afm.py:75]: Same comment.

**Review by lamoreel — CHANGES_REQUESTED** [review]
We should remember that this strategy can also be used to make similar tests with arbitrary DMI tensor elements (so no special cases such as `bulk_dmi`).

**Review by lamoreel — APPROVED** [review]

**Review by ilateur — APPROVED** [review]
